### PR TITLE
[Serializer] Add a 'make:serializer:normalizer' command

### LINKED ISSUE #287: Added a command to create a Normalizer
> I think making custom normalizer is much more likely to happen than custom encoders though.

_Originally posted by @stof in https://github.com/symfony/maker-bundle/pull/30#issuecomment-347550483_

---

I totally agree. Of course it would be stupid to remove the `make:serializer:encoder` command. 

But a `make:serializer:normalizer` would be super useful

### PR DESCRIPTION
**Pull request by lyrixx**
fixes #287

### CODE REVIEW

File: src/Resources/skeleton/serializer/Normalizer.tpl.php [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+        $data = $this->normalizer->normalize($object, $format, $context);
+
+        // Here: add, edit, or delete some data
+
+        return $data;
+    }
+
+    public function supportsNormalization($data, $format = null): bool
+    {
+        return $data instanceof \App\Entity\BlogPost;
```
Comment (franek) [inline src/Resources/skeleton/serializer/Normalizer.tpl.php]: I think you should update this line with $namespace/$class_name ?

Comment (lyrixx) [reply src/Resources/skeleton/serializer/Normalizer.tpl.php]: `$namespace\$class_name` === `App\Serializer\Normalizer\TopicNormalizer` (for exemple)

So no. Did I miss something?

File: src/Maker/MakeSerializerNormalizer.php [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        ;
+    }
+
+    public function generate(InputInterface $input, ConsoleStyle $io, Generator $generator)
+    {
+        $normalizerClassNameDetails = $generator->createClassNameDetails(
+            $input->getArgument('name'),
+            'Serializer\\Normalizer\\',
+            'Normalizer'
+        );
```
Comment (weaverryan) [inline src/Maker/MakeSerializerNormalizer.php]: For the serializer encoder maker, we just put it into the `Serializer` directory (not a sub-directory). I'm interested in how many encoders / normalizers applications have, and whether we should put them in sub-directories, or just in Serializer (to keep things a bit smaller).

File: src/Resources/help/MakeSerializerNormalizer.txt [hunk 3]
```diff
@@ -0,0 +1,5 @@
+The <info>%command.name%</info> command generates a new serializer normalizer class.
+
+<info>php %command.full_name% UserEncoder</info>
```
Comment (weaverryan) [inline src/Resources/help/MakeSerializerNormalizer.txt]: UserNormalizer?

File: src/Maker/MakeSerializerNormalizer.php (hunk 2 above)
Comment (lyrixx) [reply src/Maker/MakeSerializerNormalizer.php]: I know and I don't like it.
I have the same opinion about form type.

It really prefer to use a sub namespace to keep things clean. 

I usually have between 0 to 10 normalizers and sometimes 1 encoder. 

Btw voters are put in a sub namespace too. 

It's just an extra directory...

**Comment by lyrixx** [comment]
Hello, 

> But if you're decoding, you'll need the denormalizer as well.

This is not true. de-normalization is build'in. In looked at my application, and I never had to create a de-normalizer. However, I have many normalizers.

So, I'm not sure it's a goo idea to add this option

File: src/Resources/help/MakeSerializerNormalizer.txt (hunk 3 above)
Comment (lyrixx) [reply src/Resources/help/MakeSerializerNormalizer.txt]: good catch. Fixed. Thanks

File: src/Resources/skeleton/serializer/Normalizer.tpl.php (hunk 1 above)
Comment (franek) [reply src/Resources/skeleton/serializer/Normalizer.tpl.php]: No. I read your PR too quickly ;) ok for me ;)

**Comment by lyrixx** [comment]
Hello, Could we merge this PR? It's really helpful during training an real life.
Regarding the deserialize option, since my last comment, I did not have to create a denormalizer ... So I'm still against this option

**Comment by weaverryan** [comment]
Thank you Grégoire! If you want, you could update the docs to mention the new command :)

**Comment by dunglas** [comment]
It could be nice to make the generated normalizer implementing the `CacheableSupportsMethodInterface`. Almost all normalizers will benefit from it in term of performance.

File: src/Resources/skeleton/serializer/Normalizer.tpl.php (hunk 1 above)
Comment (OskarStark) [inline src/Resources/skeleton/serializer/Normalizer.tpl.php]: @lyrixx why are you checking against `App\Entity\Blogpost`? 🤔

File: src/Resources/skeleton/serializer/Normalizer.tpl.php [hunk 4]
```diff
@@ -0,0 +1,30 @@
+<?= "<?php\n" ?>
+
+namespace <?= $namespace; ?>;
+
+use Symfony\Component\Serializer\Normalizer\NormalizerInterface;
+use Symfony\Component\Serializer\Normalizer\ObjectNormalizer;
+
+class <?= $class_name ?> implements NormalizerInterface
```
Comment (OskarStark) [inline src/Resources/skeleton/serializer/Normalizer.tpl.php]: Shouldn't this be `final`?

File: src/Resources/skeleton/serializer/Normalizer.tpl.php (hunk 1 above)
Comment (lyrixx) [reply src/Resources/skeleton/serializer/Normalizer.tpl.php]: Why not ? it's just an exemple, as in Voter

File: src/Resources/skeleton/serializer/Normalizer.tpl.php (hunk 4 above)
Comment (lyrixx) [reply src/Resources/skeleton/serializer/Normalizer.tpl.php]: It could indeed.Voter are not. I was consistant

**Pull request merged**